File: Python/Nonlineq/BisNewtonConvergenceExer5P5b.py

import numpy as np #Basic linear algebra
import matplotlib.pyplot as plt #Plotting

tol = 10*10**(-16)
def f(x):
    return np.exp(-x) -x
def f_prime(x):
    return -1*np.exp(-x) -1

t=np.linspace(-1,1,50)
b=f(t)

y0=0.5 #första gissning 
y=y0-f(y0)/f_prime(y0)
newton_x=[]
newton=[]
newton_x.append(y0); newton_x.append(y)
newton.append(f(y0)); newton.append(f(y))
k3=2

while abs(y-y0)>tol:
    y0=y
    y=y0-f(y0)/f_prime(y0)
    newton_x.append(y)
    newton.append(f(y))
    k3 += 1

i3=np.linspace(1,k3,k3) #stara från 1 iteration
plt.figure(0)
plt.plot(i3,newton,"c*-",label="Newtons metod")
plt.xlabel('iteration')
plt.ylabel('function e**(-x)-x=0')
plt.title('Newtons metod, konvergens (iterations):' f'{k3}')
plt.legend()
plt.grid()


n = -1.0
p = 1.0
x_exact = np.linspace(n,p,100);
func_exact=np.ones(len(x_exact))
N = len(x_exact)
for i in range(0,N):
    func_exact[i] =f(x_exact[i])
plt.figure(1)
plt.plot(x_exact,func_exact,"b-",label="e**(-x)-x=0")
plt.xlabel('x')
plt.ylabel('function e**(-x)-x=0')
plt.title('Exact function e**(-x)-x')
plt.legend()
plt.grid()

File: MATLAB/Nonlineq/BisNewtonConvergenceExer5P5b.m
close all;

tolerance = 10e-15;

%fun = @(x)x^10 - 10^10;
%fun = @(x)x^3 - 2*x - 5;
fun =  @(x)exp(-x) -x;


% Newtons metod

y=0;
iteration_newton = 0;
func_newton = 0;
error_newton = 0;

%Initialiseringen: init guess
y(1) = 1.5
%y(1) = 7;
k=2;

iteration_newton(1) = 1;
func_newton(1) =fun(y(1));
%error_newton(1) = abs(10.0 - func_newton(1));
numerator =  fun(y(1));
denominator =  -exp(-y(1))-1;
y(2) = y(1) - numerator/denominator;
  
% Main Newton's iterations  
while abs( y(k) - y(k-1) ) > tolerance

numerator =  fun(y(k));
denominator = -exp(-y(k))-1;
y(k+1) = y(k) - numerator/denominator;
iteration_newton(k) = k;
func_newton(k) =fun(y(k));

k = k+1;
  end


figure
 % plot(iteration_newton,error_newton, 'o b-','LineWidth',2)
 plot(iteration_newton,func_newton, 'o b-','LineWidth',2)
  
  xlabel('iteration')
  ylabel('function   e^{-x} - x = 0')
grid on
 legend(['solution:',num2str(y(k))]);
 title(['Newtonsmetod, konvergens (iterations):',num2str(k)])

n= -1.0;
p=3.0;
x_exact = linspace(n, p);
N = size(x_exact,2);
 for i = 1:N
	    func_exact(i) =fun(x_exact(i));
end

figure

plot(x_exact,func_exact, ' b-','LineWidth',2)
 legend('  e^{-x} - x = 0');

  xlabel('  x')
  ylabel('function  e^{-x} - x = 0')

grid on
iteration


 iteration_newton
